packages/@expo/cli/src/start/server/metro/inspector-proxy/proxy.ts

```diff
@@ -0,0 +1,145 @@
+    } else {
+      this.metroProxy._serverAddressWithPort = `localhost:${port}`;
+    }
+  }
+
+  /** @see https://chromedevtools.github.io/devtools-protocol/#endpoints */
+  public httpEndpointMiddleware(
+    req: IncomingMessage,
+    res: ServerResponse,
+    next: (error?: Error) => any
+  ) {
+    return this.metroProxy.processRequest(req, res, next);
+  }
+
+  public createWebSocketEndpoints(port = 1111): Record<string, WSServer> {
+    // Auto-initialize the server address on a best-guess basis
+    this.setServerAddress({ port, family: 'IPv4' });
+
+    return {
+      [WS_DEVICE_URL]: this.createDeviceWebSocketServer(),
+      [WS_DEBUGGER_URL]: this.createDebuggerWebSocketServer(),
+    };
+  }
+
+  private createDeviceWebSocketServer() {
+    const WS = importMetroInspectorWebSocketFromProject(this.metroProxy._projectRoot);
+    const wss = new WS.Server({
+      noServer: true,
+      perMessageDeflate: false,
+    });
+
+    // See: https://github.com/facebook/metro/blob/eeb211fdcfdcb9e7f8a51721bd0f48bc7d0d211f/packages/metro-inspector-proxy/src/InspectorProxy.js#L157
+    wss.on('connection', (socket, request) => {
+      try {
+        const deviceId = this.metroProxy._deviceCounter++;
+        const { deviceName, appName } = getNewDeviceInfo(request.url);
+
+        this.devices.set(
+          deviceId,
+          new this.DeviceClass(deviceId, deviceName, appName, socket, this.metroProxy._projectRoot)
+        );
+
+        debug('New device connected: device=%s, app=%s', deviceName, appName);
+
+        socket.on('close', () => {
+          this.devices.delete(deviceId);
+          debug('Device connected: device=%s, app=%s', deviceName, appName);
+        });
+      } catch (error: unknown) {
+        const message = error instanceof Error && error.toString();
+        console.error('Could not establish a connection to device:', error);
+        socket.close(WS_GENERIC_ERROR_STATUS, message || 'Unknown error');
+      }
+    });
+
+    return wss;
+  }
+
+  private createDebuggerWebSocketServer() {
+    const WS = importMetroInspectorWebSocketFromProject(this.metroProxy._projectRoot);
+    const wss = new WS.Server({
+      noServer: true,
+      perMessageDeflate: false,
+    });
+
+    // See: https://github.com/facebook/metro/blob/eeb211fdcfdcb9e7f8a51721bd0f48bc7d0d211f/packages/metro-inspector-proxy/src/InspectorProxy.js#L193
+    wss.on('connection', (socket, request) => {
+      try {
+        const { deviceId, pageId } = getExistingDeviceInfo(request.url);
+        if (!deviceId || !pageId) {
+          throw new Error(`Missing "device" and/or "page" IDs in query parameters`);
+        }
+
+        const device = this.devices.get(parseInt(deviceId, 10));
+        if (!device) {
+          throw new Error(`Device with ID "${deviceId}" not found.`);
+        }
+
+        debug('New debugger connected: device=%s, app=%s', device._name, device._app);
+
+        device.handleDebuggerConnection(socket, pageId);
```

Comment:
Add a `socket.on('close')` to add a debug log whenever a debugger disconnects. E.g.

```suggestion
        device.handleDebuggerConnection(socket, pageId);
        socket.on('close', () => {
          debug('Debugger connected: device=%s, app=%s', device._name, device._app);
        });
```